### PR TITLE
[FIX] website_sale: only show strikethrough price for discount rules

### PR DESCRIPTION
**Pull request by mash-odoo**
**Steps:**
- Create a new pricelist
- Apply pricelist rules
- For eg: Apply it on Acoustic Bloc Screen, with a fixed price of 20.
- When seen on the shop page or product page, it shows a strikethrough price
along with the original price
- Same happens when we use formula instead of fixed price.

**Changes:**
Changed some test cases because currently, the pricelist had fixed price
which was causing the strikethrough to not appear. But due to the code
change, it will now not appear, hence giving the percentage price to the
tour to run it smoothly.

**Before this commit:**
The strikethrough prices on the shop page were shown for all the pricelist rules

**After this commit:**
The strikethrough prices are now shown everywhere only if the pricelist rule is a 'discount'.
Formula based discount rules strikethrough prices are also displayed on the shop/configurator and product pages, but not on the cart and subsequent steps of the checkout to allow customer to apply discounts while still keeping 'beautiful' prices.

**Affected version:** saas-17.4~master
opw-4181825

### CODE REVIEW

File: addons/website_sale/models/product_template.py [hunk 1]
```diff
@@ -468,6 +470,9 @@ def _get_additionnal_combination_info(self, product_or_template, quantity, date,
         )
 
         price_before_discount = pricelist_price
+        pricelist_item = self.env['product.pricelist.item'].browse(pricelist_rule_id)
+        pricelist_compute_price = pricelist_item.compute_price
```
Comment (kcv-odoo) [inline addons/website_sale/models/product_template.py:474]: should be inside if condition
Also no need for variable `pricelist_compute_price`

File: addons/website_sale/models/product_template.py [hunk 2]
```diff
@@ -287,14 +287,16 @@ def _get_sales_prices(self, website):
...
+                pricelist_compute_price = pricelist_item.compute_price
                 pricelist_base_price = template.env['product.pricelist.item']._compute_base_price(
                     product=template,
                     quantity=1.0,
                     date=date,
                     uom=template.uom_id,
                     currency=currency,
                 )
-                if pricelist_base_price != pricelist_price:
+                if pricelist_base_price != pricelist_price and pricelist_compute_price == "percentage":
```
Comment (kcv-odoo) [inline addons/website_sale/models/product_template.py:299]: why we only want to display this for `percentage` price not for other `formula` and `fixed` price? :thinking:

**Review by kcv-odoo — CHANGES_REQUESTED** [review]
Thanks for the fix!

**Comment by kcv-odoo** [comment]
FYI runbot is also red because of you changes https://runbot.odoo.com/runbot/build/68564663 :slightly_frowning_face:

File: addons/website_sale/models/product_template.py (hunk 2 above)
Comment (mash-odoo) [reply addons/website_sale/models/product_template.py:299]: @dalaOdoo mentioned on the task and suggested that the strikethrough should only be `applied for discount` and `not for fixed or formula`

**Comment by mash-odoo** [comment]
Yes working on the test case

File: addons/website_sale/models/product_template.py (hunk 2 above)
Comment (dalaOdoo) [reply addons/website_sale/models/product_template.py:299]: The biggest problem we need to solve is the fact that currently a pricelist in another currency behave as a discount with the strike through price.

File: addons/website_sale/tests/test_website_sale_show_compare_list_price.py [hunk 3]
```diff
@@ -75,8 +75,8 @@ def setUpClass(cls):
                 Command.create({
                     'applied_on': '1_product',
                     'product_tmpl_id': cls.test_product_with_pricelist_and_compare_list_price.id,
-                    'compute_price': 'fixed',
-                    'fixed_price': 3500,
+                    'compute_price': 'percentage',
```
Comment (kcv-odoo) [inline addons/website_sale/tests/test_website_sale_show_compare_list_price.py:78]: why do we need to change test cases? :thinking:

Comment (mash-odoo) [reply addons/website_sale/tests/test_website_sale_show_compare_list_price.py:78]: These test cases have fixed prices which is checked in the tour to trigger and click, but due to our specification, we now only show strikethrough when compute_price is percentage. Hence for the tour we had 2 approaches:
1. Either remove the part where we checked the base_price as it will not be shown when we use fixed price or formula
2. Adapt the tour accordingly where we show percentage price which will be visible for strikethrough.

File: addons/website_sale/models/product_template.py [hunk 4]
```diff
@@ -471,6 +475,7 @@ def _get_additionnal_combination_info(self, product_or_template, quantity, date,
         if pricelist_rule_id:  # If a rule was applied, there might be a discount
             # For ecommerce flows, the base price is always the product sales price
             # which can be computed by calling `_compute_base_price` without a pricelist rule
+            pricelist_item = self.env['product.pricelist.item'].browse(pricelist_rule_id)
```
Comment (kcv-odoo) [inline addons/website_sale/models/product_template.py:478]: should initialize `pricelist_item` variable before if to avoid error when there in no `pricelist_rule_id`

pricelist_item = self.env['product.pricelist.item']

File: addons/website_sale/tests/test_website_sale_show_compare_list_price.py (hunk 3 above)
Comment (kcv-odoo) [reply addons/website_sale/tests/test_website_sale_show_compare_list_price.py:78]: IMO we should have test cases for both flows
- for existing ``fixed` compute price adept tour 
- Add another tour for `percentage`  with strike-through

To ensure we are not breaking fix compute price flow

File: addons/website_sale/models/product_template.py [hunk 5]
```diff
@@ -287,14 +287,18 @@ def _get_sales_prices(self, website):
...
                     product=template,
                     quantity=1.0,
                     date=date,
                     uom=template.uom_id,
                     currency=currency,
                 )
-                if pricelist_base_price != pricelist_price:
+                if (
+                    pricelist_base_price != pricelist_price
+                    and pricelist_item.compute_price == 'percentage'
```
Comment (Feyensv) [inline addons/website_sale/models/product_template.py:300]: Don't compute the base price if the pricelist rule is not a percentage one.

File: addons/website_sale/models/product_template.py [hunk 6]
```diff
@@ -471,6 +475,7 @@ def _get_additionnal_combination_info(self, product_or_template, quantity, date,
         if pricelist_rule_id:  # If a rule was applied, there might be a discount
             # For ecommerce flows, the base price is always the product sales price
             # which can be computed by calling `_compute_base_price` without a pricelist rule
+            pricelist_item = self.env['product.pricelist.item'].browse(pricelist_rule_id)
             price_before_discount = self.env['product.pricelist.item']._compute_base_price(
```
Comment (Feyensv) [inline addons/website_sale/models/product_template.py:479]: same here

File: addons/website_sale/tests/test_website_sale_show_compare_list_price.py (hunk 3 above)
Comment (Feyensv) [reply addons/website_sale/tests/test_website_sale_show_compare_list_price.py:78]: Don't spend too much time on adding new tests here, adapting them is good already for now, unless you really notice an issue that wasn't caught by the test when it should have.
Since this is a functional change (even if it makes sense), it should be deployed quickly to avoid users becoming used to the current behavior.

**Comment by Feyensv** [comment]
Also, forgot to say, but the product configurator must be updated too to display the right strikethrough price

**Comment by Feyensv** [comment]
Hello @Jean-boje, can you take a look and tell us if you believe we really need to change that in stable (DALA thinks so :D) ?

We changed the pricelists logic a lot in 17.4, but the first version of the work didn't seem safe enough for ecommerce (and PAJU was absent), so we didn't dare to change this logic in ecommerce.  Later on (still in 17.4), we changed to a second approach where discounts (and strikethrough prices) are only displayed when the pricelist rules is a 'percentage' discount.
Until now, we only applied that logic in the backend sales logic, but we believe it now makes sense to harmonize the behavior with ecommerce as well.  It makes sense to do so in 17.4, but it's still a functional change and some existing databases would see direct changes to their displayed discounts if we merge this PR.

**Comment by Jean-boje** [comment]
Hi @Feyensv, so this change would force that Discount pricelists be shown as Discounts when a Formula isn't? 
So, if I want to apply a clean X,99€ price for my rebate, will it never show the discounted price? 
What do you think of adding a checkbox under the E-commerce tab to precise if we want to display or not a strikethrough price for this specific pricelist?
IMO, this my solve the issue without forcing a not 100% logical behavior.

File: addons/website_sale/models/product_template.py [hunk 7]
```diff
@@ -284,10 +284,9 @@ def _get_sales_prices(self, website):
                     pricelist_price, currency, product_taxes, taxes, self, website=website,
                 ),
             }
-            if pricelist_rule_id:  # If a rule was applied, there might be a discount
-                # For ecommerce flows, the base price is always the product sales price
-                # which can be computed by calling `_compute_base_price` without a pricelist rule
-                pricelist_base_price = template.env['product.pricelist.item']._compute_base_price(
+            pricelist_item = template.env['product.pricelist.item'].browse(pricelist_rule_id)
+            if pricelist_rule_id and pricelist_item.compute_price == 'percentage':
+                pricelist_base_price = pricelist_item._compute_base_price(
```
Comment (ltinel) [inline addons/website_sale/models/product_template.py:289]: Was this change intentional? `_compute_base_price` was called on an empty recordset before.

**Review by ltinel — CHANGES_REQUESTED** [review]
Thanks for your fix! :slightly_smiling_face:

File: addons/website_sale/models/product_template.py (hunk 7 above)
Comment (Feyensv) [reply addons/website_sale/models/product_template.py:289]: Yes, previously the base price on ecommerce was always the sales price, whereas in sales it was the pricelist base price.
As we want to harmonise ecommerce and sales behavior, we'll compute the strikethrough price as the pricelist rule base price, and not always the sales price anymore.

File: addons/sale/controllers/product_configurator.py [hunk 8]
```diff
@@ -362,14 +368,16 @@ def _get_basic_product_information(self, product_or_template, pricelist, combina
...
         return dict(
             **basic_information,
-            price=pricelist._get_product_price(
-                product_or_template.with_context(
-                    **product_or_template._get_product_price_context(combination)
-                ),
-                **kwargs,
-            ),
+            price=price,
+            pricelist_rule_id=pricelist_rule_id,
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/controllers/product_configurator.py:380]: I did that part of the code, feel free to push another version, but I felt like this was the easiest approach without polluting the code (and without much api changes in stable).

File: addons/website_sale/models/product_template.py (hunk 7 above)
Comment (ltinel) [reply addons/website_sale/models/product_template.py:289]: Hmm, should we do the same below then? (in `_get_additionnal_combination_info`)

Comment (Feyensv) [reply addons/website_sale/models/product_template.py:289]: Indeed, I thought I did it but it seems I forgot :D

File: addons/website_sale/controllers/product_configurator.py [hunk 9]
```diff
@@ -251,20 +255,24 @@ def _get_strikethrough_price(self, product_or_template, currency, date, price):
...
-        )
-        if currency.compare_amounts(sales_price, price) == 1:
-            # apply taxes
-            return self._apply_taxes_to_price(
-                sales_price,
+        pricelist_rule = request.env['product.pricelist.item'].browse(pricelist_rule_id)
+
+        # Only show strikethrough if compute_price is set to 'percentage'
+        if pricelist_rule and pricelist_rule.compute_price == 'percentage':
+            sales_price = request.env['product.pricelist.item']._compute_base_price(
```
Comment (Feyensv) [inline addons/website_sale/controllers/product_configurator.py:262]:
```suggestion
            sales_price = pricelist_rule._compute_base_price(
```

**Review by ltinel — APPROVED** [review]
LGTM :)

**Comment by Feyensv** [comment]
@robodoo r+ squash
